### PR TITLE
Introduce corpus functions

### PR DESCRIPTION
**Pull request by mdepinet**
This is an alternative to https://github.com/fixie-ai/fixie-sdk/pull/178

### CODE REVIEW

File: fixieai/agents/agent_func.py [hunk 1]
```diff
@@ -232,26 +307,93 @@ def create(
...
+        ):
+            raise TypeError(
+                f"Registered function {func_name} cannot accept variable args: {params!r}."
+            )
+
+        if len(params) > 1:
+            raise TypeError(
+                f"Registered function {func_name} must have one or zero arguments."
+            )
+        param = next(iter(params.keys())) if params else None
```
Comment (petersalas) [inline fixieai/agents/agent_func.py:369]:
```suggestion
        param = next(iter(params.keys()), None)
```

File: fixieai/agents/agent_func.py [hunk 2]
```diff
@@ -232,26 +307,93 @@ def create(
...
+                [
+                    (corpora.CorpusRequest, lambda request, claims: request),
+                    (param, lambda request, claims: request),
+                ],
+            )
+            if param
+            else None
+        )
+
+        if bool(params) != bool(bound_request_mapper):
```
Comment (petersalas) [inline fixieai/agents/agent_func.py:383]: A perhaps-simpler alternative to going through the `_bind_argument_mapper` machinery could be to directly validate the type annotation of the parameter, but consistency with the other function is fine too.

File: fixieai/agents/corpora.py [hunk 3]
```diff
@@ -1,18 +1,147 @@
...
 from typing import List, Optional
 
+import dataclasses_json.core
 from pydantic import dataclasses as pydantic_dataclasses
 
 
 @pydantic_dataclasses.dataclass
-class DocumentLoader:
-    """Document loader for a Fixie CodeShot Agent."""
+class CorpusRequest(dataclasses_json.DataClassJsonMixin):
```
Comment (petersalas) [inline fixieai/agents/corpora.py:10]: Removing `DocumentLoader` seems like a breaking change? Would recommend leaving it for now unless we need to remove it.

File: fixieai/agents/corpora.py [hunk 4]
```diff
@@ -22,4 +151,6 @@ class DocumentCorpus:
     """A list of wildcard patterns to exclude from crawled URLs (e.g., */no_crawl/*)."""
 
     auth_token_func: Optional[str] = None
-    loader: Optional[DocumentLoader] = None
+
+
+DocumentCorpus = UrlDocumentCorpus  # Deprecated: Prefer UrlDocumentCorpus
```
Comment (petersalas) [inline fixieai/agents/corpora.py:156]: Is there a way to specify the new corpus type?

File: fixieai/agents/agent_func.py [hunk 5]
```diff
@@ -232,26 +307,93 @@ def create(
...
+        if isinstance(result, CorpusResponse):
+            yield result
+        else:
+            raise TypeError(
+                f"The func result was type {type(result)}, but must be a CorpusResponse."
+            )
+
+    def adapt_exception(self, e: exceptions.AgentException) -> CorpusResponse:
+        logging.warn(f"Corpus loading threw an exception: {e}", stack_info=True)
+        return corpora.CorpusResponse()
```
Comment (petersalas) [inline fixieai/agents/agent_func.py:335]: I think this should probably just raise and HTTP 500 -- unlike `AgentQuery`/`AgentReponse` there's not a human or LLM on the other end of the request.

File: fixieai/agents/agent_base.py [hunk 6]
```diff
@@ -195,6 +251,45 @@ def _serve_func(
...
+                http_status_code=404,
+                func_name=corpus_func_name,
+                available_funcs=list(self._corpus_funcs.keys()),
+            )
+
+        output = pyfunc(request, token_claims)
+
+        return next(iter(output))
+
+    def _check_credentials(
```
Comment (petersalas) [inline fixieai/agents/agent_base.py:283]: might as well use this in `standalone.py` as well

**Review by petersalas — APPROVED** [review]
I like it! Curious what @juberti's take is as well.

File: fixieai/agents/corpora.py (hunk 3 above)
Comment (mdepinet) [reply fixieai/agents/corpora.py:11]: AFAICT, this never did anything so I figured we could bump the version to be safe and remove it.  For now I've added it back and deprecated it instead.

File: fixieai/agents/corpora.py (hunk 4 above)
Comment (mdepinet) [reply fixieai/agents/corpora.py:175]: Whoops, I forgot to add the registered function to the agent metadata. 😅 Added.

File: fixieai/agents/agent_func.py (hunk 2 above)
Comment (mdepinet) [reply fixieai/agents/agent_func.py]: I initially wasn't sure this was going to be simpler, but I think you're right. Done.

**Pull request merged**